### ISSUE TITLE
Multi-threaded configuration of pipeline executables

### ISSUE THREAD

**Issue by maak** [openMVG/openMVG#420]
I have found that the openMVG_main_ComputeMatches application runs on a single thread. Understandably, this takes a very long time (days on a macbook pro). It would be useful to include an option for specifying a threadpool size to be utilized by the process.

**Comment by maak** [comment]
I am not familiar with OpenMP. I see it is mentioned in make files. Perhaps this functionality is already available? I will try setting OMP_NUM_THREADS to something other than 1

**Comment by maak** [comment]
I tried setting OMP_NUM_THREADS to 8. Unfortunately that didn't have an impact on the thread utilization.

**Comment by pmoulon** [comment]
HI @maak,
OpenMVG use OpenMP for threading.
I think you are on a mac with clang and that your clang does not include OpenMP!
Please check it (cmake must write if OpenMP is detected or not).
I think that the last update of clang have perhaps an OpenMP support or another solution can be to compile OpenMVG with GCC.

PS: Some given module will certainly use std::thread later on for doing MT task rather than using OpenMP.

**Comment by maak** [comment]
You are correct, that is the issue. I tried compiling with a clang variant of OpenMP (https://clang-omp.github.io/), but I do not observe multiple threads running during putatative matches.

I am trying to compile using gcc (by setting DCMAKE_C_COMPILER and DCMAKE_CXX_COMPILER to gcc variants), but I get a lot of errors similar to this while compiling the doc target:

/var/folders/h0/9s9n7qmx6wxbp4z4sf7tcp1c0000gp/T//ccSjqnPQ.s:360:no such instruction: `vmovsd 16(%rbp), %xmm0'

Any ideas, or is this a mac-specific problem?

(Thanks for the help, by the way)

**Comment by simogasp** [comment]
Hi,
I'm using that version of clang too and I have no problems, all the programs are multithreaded.
Are u passing  DCMAKE_C_COMPILER and DCMAKE_CXX_COMPILER with the path to ur multithreaded version of clang?
For example in my case I always run cmake with
```
cmake ../src -DCMAKE_C_COMPILER=~/dev/code/clang/buildcmake/install/bin/clang  -DCMAKE_CXX_COMPILER=~/dev/code/clang/buildcmake/install/bin/clang++ 
```
S.

**Comment by maak** [comment]
When building with clang I was not specifying a CMAKE_C(XX) compiler, so I presume it was falling back to the system clang version. 

You mention a multithreaded version of clang. I am using the following:

Apple LLVM version 5.1 (clang-503.0.38) (based on LLVM 3.4svn)
Target: x86_64-apple-darwin14.5.0
Thread model: posix

I believe this is the default installation. Is there a variant with additional multi-threading support available? If so, how does openMVG utilize this functionality? From what PM said above, openMVG relies solely on OpenMP for parallel processing (I kinda like OpenMP now that I have read more about it).

**Comment by simogasp** [comment]
I built myself the clang version supporting multithreading with OpenMP, https://clang-omp.github.io/ as you said.
Then when you run cmake you have to pass the path to the executables you have built, ie in my case I built them in `~/dev/code/clang/buildcmake/install/`, and i need to run cmake with this line 
```
cmake ../src -DCMAKE_C_COMPILER=~/dev/code/clang/buildcmake/install/bin/clang  -DCMAKE_CXX_COMPILER=~/dev/code/clang/buildcmake/install/bin/clang++ 
```

As for the version, from what you posted it seems to me that you ran a `gcc -v`, that will return the version of the default version of clang because gcc is a default link to the default compile, my `gcc -v` reads indeed
```
Apple LLVM version 5.1 (clang-503.0.40) (based on LLVM 3.4svn)
Target: x86_64-apple-darwin13.3.0
Thread model: posix
```
while if I run `~/dev/code/clang/buildcmake/install/bin/clang -v` I get 
```
clang version 3.5.0 
Target: x86_64-apple-darwin13.3.0
Thread model: posix
```
So long story short, just run cmake with those flags and everything should work.

**Comment by simogasp** [comment]
Ah, I forgot, there is a little trick, you won't be able to compile Ceres library as it is with the OpenMP clang because the cmake needs some adjustments to make it work with that version. So the workaround is to download Ceres separately and compile it with the default clang, and then use that version when you compile OpenMVG by passing 
`-DCeres_DIR=/path/to/ceres/install/share/Ceres`
to cmake.

**Comment by rperrot** [comment]
I used to compile openMVG using clang-omp but you have to check in CMakeLists.txt (don't know exactly the good one) to remove check of clang (with disable openMP), this is why even using clang-omp you don't have openmp activated.

**Comment by igorti** [comment]
Another way to enable OpenMP on Mac is to use `gcc` when building OpenMVG. Then it's just a matter of passing right flags before compiling:

```
  export CC="/usr/local/bin/gcc-5"
  export CXX="/usr/local/bin/c++-5"
  export CPPFLAGS='-I/usr/local/include/'
  export LDFLAGS='-L/usr/local/lib/'
```

Notice though, that openMVG_main_ComputeFeatures is single threaded, so that won't help much anyway, but other steps in pipeline run much faster with OpenMP.

**Comment by rperrot** [comment]
You need to install gcc5 from macports or homebrew (or by hand). 

The lack of multithreading in ComputeFeatures, is due to vlfeat, not directly OpenMVG related. If you're using small to medium size pictures it wouldn't be such a problem, since mainly matching is the main bottleneck in the whole process. 

Moreover, with AKAZE descriptor, features computation can exploit multithreading.

**Comment by igorti** [comment]
Interesting! Anything specific you have to do to exploit multithreading? I'm always using AKAZE_FLOAT descriptor and it seems like ComputeFeatures is single threaded all the way.

**Comment by rperrot** [comment]
It's strange, because if you have OpenMP enabled, it should be done in multithread way. 

For example feature refinement is done with OpenMP when available.  

https://github.com/openMVG/openMVG/blob/master/src/openMVG/features/akaze/AKAZE.cpp#L496

Could you check if OPENMVG_USE_OPENMP is set ?

**Comment by igorti** [comment]
Yes, it's definitely set - I can see it is ON during  compilation, then I also tried conditional output with this flag during the execution and it works. Now that I look closer, I can see on the CPU usage that there are occasional spikes when several threads are used, but most of the time it's utilising only one thread. 

Probably parallelisation should be done on higher level to really benefit from it, like suggested in [#490](https://github.com/openMVG/openMVG/pull/490)?

**Comment by pmoulon** [comment]
@igorti Existing parallelization for SIFT is not good actually (since it compute in parallel only the descriptor extraction step).
I have tried in the past the suggestion from #490 and it was not producing the same files on various run.
@mitjap have made a second try and this time it seems to work. (Option that try to use many images at the same time can be dangerous, since it can use more ram than the machine have and crash directly...). But for experimented users it's ok.

The CPU usage when you are using AKAZE_FLOAT must be visible...

@igorti Would you be ok with the fact to test the #490 in order to check:
- that the same SIFT are computed (as with one thread)?
- that you have a speed gain too?

**Comment by igorti** [comment]
I did a few tests on a small dataset with and without #490 and here are my findings:

 - The md5-checksums of `*.feat` and `*.desc` files are different. BUT - they are also different on every new run with a single threaded version. I manually checked few files and they are mostly the same with some difference in a few places.

 - Checksums of `sfm_data.json` and `image_describer.json` files are identical with/without omp.

 - I ran through the whole pipeline with SIFT/AKAZE descriptors and single/multithreaded versions and the output is pretty much the same.

- Most importantly - the speed improvement is dramatic. On HIGH describerPreset multithreaded version was twice as fast(7 vs 14 mins) and I guess on large datasets the difference will be even bigger.

- Memory usage grows with the number of threads though. While single threaded version was using roughly 1GB of memory, running 8 threads would take 8GB. So it comes at a cost, but it's well worth the time savings.
